test/adt_test.exs

```diff
@@ -61,4 +61,12 @@ defmodule AdtTest do
       _ -> false
     end)
   end
+
+  test "adtcase allows per-module pattern matching" do
+    foo = %AdtDefinition.Foo{}
+    assert 0 == (ADT.case AdtDefinition, foo do
```

Comment:
gave that a shot here: https://github.com/vendethiel/ADT.ex/pull/6

Definitely needs some thinking to make it nice to use and also maintainable!